src/D2L.Security.OAuth2/Keys/KeyManagementService.cs

```diff
@@ -0,0 +1,207 @@
+﻿using D2L.Security.OAuth2.Keys.Default;
+using D2L.Security.OAuth2.Utilities;
+using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace D2L.Security.OAuth2.Keys {
+	internal sealed class KeyManagementService : IKeyManagementService, IPrivateKeyProvider, IDisposable {
+		private readonly IPublicKeyDataProvider m_publicKeys;
+		private readonly IPrivateKeyDataProvider m_privateKeys;
+		private readonly IDateTimeProvider m_clock;
+		private readonly OAuth2Configuration m_config;
+
+		private D2LSecurityToken m_current = null;
+
+		public KeyManagementService(
+			IPublicKeyDataProvider publicKeys,
+			IPrivateKeyDataProvider privateKeys,
+			IDateTimeProvider clock,
+			OAuth2Configuration config
+		) {
+			m_publicKeys = publicKeys;
+			m_privateKeys = privateKeys;
+			m_clock = clock;
+			m_config = config;
+		}
+
+		async Task<D2LSecurityToken> IPrivateKeyProvider.GetSigningCredentialsAsync() {
+			var current = Volatile.Read( ref m_current );
+
+			var now = m_clock.UtcNow;
+
+			if ( current == null || current.ValidTo <= now ) {
+				// Slow path: RefreshKeyAsync() wasn't called on boot and/or it
+				// isn't being called in a background job.
+				await RefreshKeyAsync( current, now )
+					.ConfigureAwait( false );
+
+				current = Volatile.Read( ref m_current );
+			}
+
+			return current.Ref();
+		}
+
+		async Task<TimeSpan> IKeyManagementService.RefreshKeyAsync() {
+			var current = Volatile.Read( ref m_current );
+
+			var now = m_clock.UtcNow;
+
+			await RefreshKeyAsync(
+				current,
+				now
+			).ConfigureAwait( false );
+
+			current = Volatile.Read( ref m_current );
+
+			var expectedNextRotation = current.ValidTo - m_config.KeyRotationBuffer;
+
+			if( now > expectedNextRotation ) {
+				// If that's in the past, use a short retry window.
+				return TimeSpan.FromSeconds( 10 );
+			} else {
+				// Otherwise use that but with a little buffer for key
+				// generation time/imprecisely scheduled cron jobs.
+				return expectedNextRotation.AddMinutes( 1 ) - now;

```

Comment:
Both of these numbers assume a little bit about the numbers in the config. `OAuth2Configuration` sets some defaults and minimums for these numbers so that this is "obviously" safe -- see it's class further down.